File: is-power-of-three.clj
(ns three
	(:require [clojure.math.numeric-tower :as math]))

; run with
; $ java -cp /Users/ted/.m2/repository/org/clojure/clojure/1.4.0/clojure-1.4.0.jar:/Users/ted/.m2/repository/org/clojure/math.numeric-tower/0.0.2/math.numeric-tower-0.0.2.jar clojure.main is-power-of-three.clj 

; Performance
; real	0m7.096s
; user	0m8.074s
; sys	0m0.348s

(defn divides-by-3? [x]
	(= 0 (mod x 3)))

(defn power-of-3? [candidate]
	(if (< 0 candidate) (Exception.))
	(if (= 0 candidate)
		true
		(if-not (divides-by-3? candidate)
			false
			(if (= candidate 3)
				true
				(power-of-3? (/ candidate 3))))))


(doseq [x (filter power-of-3? (range 1000000))]
	(println x "is a power of three"))

File: IsPowerOfThree.java
/**
 * Performance (time)
 * real	0m0.193s
 * user	0m0.158s
 * sys	0m0.044s
 */
public class IsPowerOfThree {
	
	public static boolean isPowerOfThree(int x) {
		if (x < 0) throw new IllegalArgumentException();
		while (x > 1) {
			if (x % 3 == 0) {
				x = x/3;
			} else {
				return false;
			}
		}
		// 0, 1 and cubes fall through to true
		return true;
	}

	public static void main(String... args) {
		for (int i = 0; i < 1000000; i++) {
			if (isPowerOfThree(i)) {
				System.out.println(i + " is a power of three");
			} 
		}	

	}

}